File: src/sandbox.ts
import { spawn } from "node:child_process";
import { chmod, lstat, mkdir, readdir, rm, writeFile } from "node:fs/promises";
import { resolve } from "node:path";
import type { Settings } from "./config.js";
import { ValidationError } from "./database.js";
import { newId } from "./serialization.js";

export interface SandboxSpec {
  runId: string;
  cpuLimit: number;
  memoryMb: number;
  diskMb: number;
  timeoutSeconds: number;
  networkEnabled: boolean;
  image?: string;
}
export interface SandboxHandle {
  id: string;
  root: string;
  provider: "process" | "docker";
  spec: SandboxSpec;
}
export interface SandboxResult {
  exit_code: number;
  stdout: string;
  stderr: string;
  timed_out: boolean;
}
const runProcess = (
  command: string,
  args: string[],
  options: {
    cwd?: string;
    timeoutMs: number;
    maxBytes?: number;
    env?: NodeJS.ProcessEnv;
  },
): Promise<SandboxResult> =>
  new Promise((resolvePromise, reject) => {
    const child = spawn(command, args, {
      cwd: options.cwd,
      env: options.env,
      stdio: ["ignore", "pipe", "pipe"],
    });
    let stdout = "",
      stderr = "",
      timed = false;
    const max = options.maxBytes ?? 2_000_000;
    child.stdout.on("data", (data) => {
      if (stdout.length < max) stdout += data.toString();
    });
    child.stderr.on("data", (data) => {
      if (stderr.length < max) stderr += data.toString();
    });
    child.on("error", reject);
    const timer = setTimeout(() => {
      timed = true;
      child.kill("SIGKILL");
    }, options.timeoutMs);
    child.on("close", (code) => {
      clearTimeout(timer);
      resolvePromise({
        exit_code: code ?? -1,
        stdout: stdout.slice(0, max),
        stderr: stderr.slice(0, max),
        timed_out: timed,
      });
    });
  });

export class SandboxService {
  constructor(private readonly settings: Settings) {}
  async create(
    spec: Partial<SandboxSpec> & { runId: string },
  ): Promise<SandboxHandle> {
    const complete: SandboxSpec = {
      cpuLimit: spec.cpuLimit ?? 1,
      memoryMb: spec.memoryMb ?? 512,
      diskMb: spec.diskMb ?? 1024,
      timeoutSeconds: spec.timeoutSeconds ?? 60,
      networkEnabled: spec.networkEnabled ?? false,
      image: spec.image ?? "node:22-alpine",
      runId: spec.runId,
    };
    const id = newId();
    const root = resolve(this.settings.sandboxRoot, id);
    await mkdir(root, { recursive: true });
    if (this.settings.sandboxProvider === "docker") await chmod(root, 0o777);
    return {
      id,
      root,
      provider: this.settings.sandboxProvider,
      spec: complete,
    };
  }
  private safe(handle: SandboxHandle, path: string) {
    const target = resolve(handle.root, path);
    if (
      !target.startsWith(`${resolve(handle.root)}/`) &&
      target !== resolve(handle.root)
    )
      throw new ValidationError("sandbox path escape blocked");
    return target;
  }
  async write(handle: SandboxHandle, path: string, data: Uint8Array | string) {
    const target = this.safe(handle, path);
    await mkdir(resolve(target, ".."), { recursive: true });
    await writeFile(target, data);
  }
  async execute(
    handle: SandboxHandle,
    command: string,
    args: string[] = [],
  ): Promise<SandboxResult> {
    if (!command || command.includes("\0"))
      throw new ValidationError("invalid command");
    const timeoutMs = handle.spec.timeoutSeconds * 1000;
    if (handle.provider === "process")
      return runProcess(command, args, {
        cwd: handle.root,
        timeoutMs,
        env: {
          PATH: process.env.PATH ?? "/usr/local/bin:/usr/bin:/bin",
          LANG: process.env.LANG ?? "C.UTF-8",
          OMOIKANE_SANDBOX: "process",
        },
      });
    const dockerArgs = [
      "run",
      "--rm",
      "--read-only",
      "--cap-drop=ALL",
      "--security-opt=no-new-privileges",
      "--user=65532:65532",
      "--pids-limit=128",
      `--cpus=${handle.spec.cpuLimit}`,
      `--memory=${handle.spec.memoryMb}m`,
      `--network=${handle.spec.networkEnabled ? "bridge" : "none"}`,
      "--tmpfs",
      `/tmp:rw,noexec,nosuid,size=64m`,
      `--mount`,
      `type=bind,source=${handle.root},target=/workspace`,
      "--workdir",
      "/workspace",
      handle.spec.image ?? "node:22-alpine",
      command,
      ...args,
    ];
    return runProcess("docker", dockerArgs, { timeoutMs });
  }
  async commandAvailable(handle: SandboxHandle, command: string) {
    if (!/^[A-Za-z0-9._+-]+$/.test(command))
      throw new ValidationError("invalid command requirement");
    const result = await this.execute(handle, "sh", [
      "-c",
      'command -v "$1" >/dev/null 2>&1',
      "omoikane-command-check",
      command,
    ]);
    return result.exit_code === 0 && !result.timed_out;
  }
  async destroy(handle: SandboxHandle) {
    const makeWritable = async (path: string): Promise<void> => {
      let info;
      try {
        info = await lstat(path);
      } catch (error) {
        if ((error as NodeJS.ErrnoException).code === "ENOENT") return;
        throw error;
      }
      if (info.isSymbolicLink()) return;
      if (info.isDirectory()) {
        await chmod(path, 0o700);
        for (const entry of await readdir(path))
          await makeWritable(resolve(path, entry));
      } else await chmod(path, 0o600);
    };
    await makeWritable(handle.root);
    await rm(handle.root, { recursive: true, force: true });
  }
}

File: src/agent_system/sandbox.py
from __future__ import annotations

import asyncio
import os
import shutil
from dataclasses import dataclass
from pathlib import Path
from typing import Protocol

from .config import Settings


@dataclass(slots=True)
class SandboxSpec:
    run_id: str
    cpu_limit: float = 1.0
    memory_mb: int = 512
    disk_mb: int = 1024
    timeout_seconds: int = 60
    network_enabled: bool = False


@dataclass(slots=True)
class SandboxHandle:
    id: str
    root: Path
    provider: str


@dataclass(slots=True)
class ExecResult:
    return_code: int
    stdout: str
    stderr: str
    timed_out: bool = False


class SandboxProvider(Protocol):
    async def create(self, spec: SandboxSpec) -> SandboxHandle: ...

    async def exec(
        self, handle: SandboxHandle, command: list[str], cwd: str = ".", timeout: int | None = None
    ) -> ExecResult: ...

    async def destroy(self, handle: SandboxHandle) -> None: ...


class LocalSandboxProvider:
    def __init__(self, settings: Settings):
        if settings.environment == "production":
            raise RuntimeError("LocalSandboxProvider is prohibited in production")
        self.root = settings.sandbox_root.expanduser().resolve()

    async def create(self, spec: SandboxSpec) -> SandboxHandle:
        root = (self.root / spec.run_id).resolve()
        if self.root not in root.parents:
            raise ValueError("invalid sandbox run id")
        root.mkdir(parents=True, exist_ok=True)
        return SandboxHandle(id=spec.run_id, root=root, provider="local")

    async def exec(
        self, handle: SandboxHandle, command: list[str], cwd: str = ".", timeout: int | None = None
    ) -> ExecResult:
        if not command or len(command) > 64:
            raise ValueError("command must contain 1-64 argv items")
        working_dir = (handle.root / cwd).resolve()
        if working_dir != handle.root and handle.root not in working_dir.parents:
            raise ValueError("sandbox cwd escapes workspace")
        working_dir.mkdir(parents=True, exist_ok=True)
        process = await asyncio.create_subprocess_exec(
            *command,
            cwd=working_dir,
            env={"PATH": os.environ.get("PATH", "/usr/bin:/bin"), "HOME": str(handle.root)},
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        try:
            stdout, stderr = await asyncio.wait_for(process.communicate(), timeout=timeout or 60)
            return ExecResult(
                return_code=process.returncode or 0,
                stdout=stdout.decode("utf-8", errors="replace")[:1_000_000],
                stderr=stderr.decode("utf-8", errors="replace")[:1_000_000],
            )
        except TimeoutError:
            process.kill()
            await process.wait()
            return ExecResult(return_code=124, stdout="", stderr="timed out", timed_out=True)

    async def destroy(self, handle: SandboxHandle) -> None:
        if handle.root.exists() and self.root in handle.root.parents:
            await asyncio.to_thread(shutil.rmtree, handle.root)


class DockerSandboxProvider:
    def __init__(self, settings: Settings, image: str = "python:3.12-slim"):
        self.root = settings.sandbox_root.expanduser().resolve()
        self.host_root = (
            settings.sandbox_host_root.expanduser().resolve()
            if settings.sandbox_host_root
            else self.root
        )
        self.image = image
        self._specs: dict[str, SandboxSpec] = {}

    async def _docker(self, *args: str, timeout: int = 60) -> tuple[int, str, str]:
        process = await asyncio.create_subprocess_exec(
            "docker",
            *args,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        stdout, stderr = await asyncio.wait_for(process.communicate(), timeout=timeout)
        return process.returncode or 0, stdout.decode().strip(), stderr.decode().strip()

    async def create(self, spec: SandboxSpec) -> SandboxHandle:
        root = (self.root / spec.run_id).resolve()
        host_root = (self.host_root / spec.run_id).resolve()
        if self.root not in root.parents or self.host_root not in host_root.parents:
            raise ValueError("invalid sandbox run id")
        root.mkdir(parents=True, exist_ok=True)
        name = f"agent-system-{spec.run_id}".replace("_", "-")[:63]
        args = [
            "run",
            "-d",
            "--name",
            name,
            "--user",
            "65534:65534",
            "--read-only",
            "--security-opt",
            "no-new-privileges",
            "--cpus",
            str(spec.cpu_limit),
            "--memory",
            f"{spec.memory_mb}m",
            "--pids-limit",
            "128",
            "--tmpfs",
            f"/tmp:rw,noexec,nosuid,size={min(spec.disk_mb, 256)}m",
            "-v",
            f"{host_root}:/workspace:rw",
            "-w",
            "/workspace",
        ]
        if not spec.network_enabled:
            args.extend(["--network", "none"])
        args.extend([self.image, "sleep", "infinity"])
        code, _, error = await self._docker(*args)
        if code != 0:
            raise RuntimeError(f"docker sandbox create failed: {error}")
        self._specs[name] = spec
        return SandboxHandle(id=name, root=root, provider="docker")

    async def exec(
        self, handle: SandboxHandle, command: list[str], cwd: str = ".", timeout: int | None = None
    ) -> ExecResult:
        if not command:
            raise ValueError("empty command")
        container_cwd = str((Path("/workspace") / cwd).as_posix())
        if ".." in Path(cwd).parts:
            raise ValueError("sandbox cwd escapes workspace")
        code, stdout, stderr = await self._docker(
            "exec", "-w", container_cwd, handle.id, *command, timeout=timeout or 60
        )
        return ExecResult(return_code=code, stdout=stdout, stderr=stderr)

    async def destroy(self, handle: SandboxHandle) -> None:
        await self._docker("rm", "-f", handle.id)
        if handle.root.exists() and self.root in handle.root.parents:
            await asyncio.to_thread(shutil.rmtree, handle.root)
